lib/pull: Add support for timestamp-check option, use in upgrader

cgwalters opened the pull request:
For both flatpak and ostree-as-host, we really want to verify up front during
pulls that we're not being downgraded. Currently both flatpak and
`OstreeSysrootUpgrader` do this before deployments, but at that point we've
already downloaded all the data, which is annoying.

Closes: https://github.com/ostreedev/ostree/issues/687

jlebon commented on tests/pull-test.sh:
```diff
@@ -35,7 +35,7 @@ function verify_initial_contents() {
 }
 
-echo "1..27"
+echo "1..28"
```
Can we also have an `ostree admin upgrade` test in `admin-test.sh` to validate the sysroot upgrader changes as well? (E.g. just checking that file objects were not fetched?).

jlebon commented on src/libostree/ostree-core.c:
```diff
@@ -2086,6 +2086,38 @@ ostree_commit_get_timestamp (GVariant  *commit_variant)
+  g_autofree char *current_ts_str = g_date_time_format (current_dt, "%c");
+  g_autofree char *new_ts_str = g_date_time_format (new_dt, "%c");
+
+  return glnx_throw (error, "Upgrade target revision '%s' with timestamp '%s' is chronologically older than current revision '%s' with timestamp '%s'; use --allow-downgrade to permit",
```
Maybe drop the `use --allow-downgrade to permit`, since we haven't necessarily gotten here through `ostree admin upgrade`, right?

Hmm, would be nice to have it appended in the sysroot upgrader instead, though we'd need to create a new error domain, it seems? (I don't see any error codes appropriate in the `G_IO_ERROR` domain at least, though maybe there's another glib domain that has something).

jlebon replied:
> Hmm, would be nice to have it appended in the sysroot upgrader instead, though we'd need to create a new error domain, it seems?

By chance, it sounds like exactly that is being discussed in https://github.com/ostreedev/ostree/issues/1064.

pwithnall replied:
I was planning on creating a special domain just for GPG-related errors; you’d probably want to create another separate one for this PR>

jlebon commented on tests/pull-test.sh:
```diff
@@ -205,6 +205,30 @@ ${CMD_PREFIX} ostree --repo=parentpullrepo rev-parse origin:main > main.txt
+fi
+assert_file_has_content err.txt "Upgrade.*is chronologically older"
+assert_streq ${newrev} "$(${CMD_PREFIX} ostree --repo=repo rev-parse main)"
+# But we can pull it without timestamp checking
```
Seems like we're missing a final `ostree pull` here without `-T` ?

cgwalters commented:
OK, rebased :surfing_man: and reworked for all comments except the error domain one; not sure I want to dive into an error domain for pulls right now.  Dropped the text in the error message referring to a command line option.

jlebon commented:
I added a fixup to simplify something in the test, which was most likely what you meant.

@rh-atomic-bot r+ 1df0034

As an experiment, I tried out using GitHub directly, which worked but it was as painful, if not more so, than dropping down the the CLI. You have to get to the file, do your edit, then Ctrl-C Ctrl-V the commit message prepended by `fixup! `.

cgwalters commented on tests/test-admin-upgrade-not-backwards.sh:
```diff
@@ -61,7 +61,7 @@ assert_file_has_content upgrade-err.txt 'chronologically older'
 assert_not_streq ${newrev} ${currev}
 assert_streq ${origrev} ${currev}
-assert_not_has_file sysroot/ostree/repo/$(ostree_checksum_to_relative_object_path sysroot/ostree/repo ${tscheck_checksum})
+assert_not_has_file sysroot/ostree/repo/$tscheck_fileobjpath
```
Hmm actually no, the repos have different modes - the first one is `.filez`, the second is `.file`.  That's why I explicitly added the code to do path to checksum and used it twice.  Right?

jlebon replied:
Ahh, sorry! Yes, that makes sense.
https://github.com/ostreedev/ostree/pull/1098
